### ISSUE TITLE
feature idea: fileline resource

### ISSUE THREAD

**Issue by mildred** [purpleidea/mgmt#187]
The idea is to have a fileline resource, like it exists in cfengine. Allows to change the content of a configuration file depending on patterns. See reference here: https://docs.cfengine.com/docs/3.5/reference-promise-types-files-edit_line.html

a list of attributes:

- select_region: the region of the file to make edits, the complete file if not specified
    - include_start_delimiter: Whether to include the section delimiter
    - include_end_delimiter: Whether to include the section delimiter
    - select_start: Regular expression matching start of edit region
    - select_end: Regular expression matches end of edit region from start
- delete_lines
    - a pattern must be provided
    - delete_if_startwith_from_list: Delete lines from a file if they begin with the sub-strings listed
    - delete_if_not_startwith_from_list: Delete lines from a file unless they start with the sub-strings in the list given
    - delete_if_match_from_list: Delete lines from a file if the lines completely match any of the anchored regular expressions listed
    - delete_if_not_match_from_list: Delete lines from a file unless the lines completely match any of the anchored regular expressions listed
    - delete_if_contains_from_list: Delete lines from a file if they contain the sub-strings listed
    - delete_if_not_contains_from_list: Delete lines from the file which do not contain the sub-strings listed
    - not_matching: When this option is true, it negates the pattern match of the promised lines
- edit_field: Certain types of text files are tabular in nature, with field separators (e.g. : or ,). The passwd and group files are classic examples of tabular files
    - allow_blank_fields: Setting allow_blank_fields defines how blank fields in a line are handled
    - extend_fields: Setting extend_fields can add new fields, to avoid triggering an error
    - field_operation: Menu option policy for editing subfields
        - append (default): Append the specified value to the end of the field/column, separating (potentially) multiple values with value_separator
        - prepend: Prepend the specified value at the beginning of the field/column, separating (potentially) multiple values with value_separator
        - alphanum: Insert the specified value into the field/column, keeping all the values (separated by value_separator) in alphanumerically sorted order
        - set: Replace the entire field/column with the specified value.
        - delete: Delete the specified value (if present) in the specified field/column.
    - field_separator: The regular expression used to separate fields in a line
    - field_value:  Set a field to a constant value
    - select_field: Sets the index of the field required 0..n (default starts from 1). 
    - start_fields_from_zero: The numbering of fields is a matter for consistency and convention. Arrays are usually thought to start with first index equal to zero (0), but the first column in a file would normally be 1. By setting this option, you can tell CFEngine that the first column should be understood as number 0 instead, for consistency with other array functions. 
    - value_separator: Defines the character separator for subfields inside the selected field
- insert_lines:
    - the lines to insert, or the file to insert lines from must be specified
    - expand_scalars: Expand any unexpanded variables
    - insert_type: Type of object the promiser string refers to
        - literal or string: Treat the promiser as a literal string of convergent lines.
        - file: The string should be interpreted as a filename from which to import lines.
        - preserve_block: The default behavior assumes that multi-line entries are not ordered specifically. They should be treated as a collection of lines of text, and not as a single unbroken object.
        - file_preserve_block: Interpret the string as a filename, and assume preserve_block semantics
    - insert_select:
        - insert_if_startwith_from_list: Insert line if it starts with a string in the list
        - insert_if_not_startwith_from_list: Insert line if it DOES NOT start with a string in the list
        - insert_if_match_from_list: Insert line if it fully matches a regex in the list
        - insert_if_not_match_from_list: Insert line if it DOES NOT fully match a regex in the list
        - insert_if_contains_from_list: Insert line if a regex in the list match a line fragment
        - insert_if_not_contains_from_list: Insert line if a regex in the list DOES NOT match a line fragment
    - location:
        - before_after: Menu option (before or after), point cursor before of after matched line
        - first_last: Choose first or last occurrence of match in file
        - select_line_matching: Regular expression for matching file line location
    - whitespace_policy: Criteria for matching and recognizing existing lines, a list of options
        - ignore_leading
        - ignore_trailing
        - ignore_embedded
        - exact_match
- replace_patterns:
    - the pattern to replace must be provided
    - occurrences: Defines which occurrences should be replaced, "all" or "first"
    - replace_value: Value used to replace regular expression matches in search

**Comment by mildred** [comment]
The main question is: do we implement that in the file resource, or do we implement that in a separate resource, considering that the output of that resource is a file.

**Comment by mildred** [comment]
Probably the most important attributes to implement are:

- a list of modifications containing:
    - select_region with include_start_delimiter, include_end_delimiter, select_start, select_end
    - delete_lines, a list of deletions:
        - pattern
    - insert_lines, a list of inserts:
        - lines: multiline string to insert
        - before_location or after_location (exclusive) with either first_pattern or last_pattern
    - replace_lines with a list of inserts:
        - pattern
        - replacement
        - occurrences: first|last

And we get the most of the feature.

**Comment by purpleidea** [comment]
@mildred I will happily ACK this feature, but have zero cycles to actually write it, so it comes with a strong "patches welcome" proviso.

The only disclaimer I have is that we think about the actual interface and name it sensibly rather than necessarily copy the cfengine interface directly. (IOW we have a chance to write this cleanly!)

> The main question is: do we implement that in the file resource, or do we implement that in a separate resource, considering that the output of that resource is a file.

I'm fairly certain it shouldn't be in the main file res. Here are a few options:

#### 1)
It could wrap the file resource and be a composite resource. See the resources/nspawn.go code for some examples of how to do this. I think this is a solid solution.

#### 2)
It could be a separate resource, but it would probably want to make use of the internal recwatch package. The downside is that we'd be re-implementing mode, user, group, etc. I think this is a bad idea.

#### 3)
It could instead be composed with two new resources.

a) A resource that reads a file's contents into an internal variable that then gets exposed to use with send->recv ( https://ttboj.wordpress.com/2016/12/07/sendrecv-in-mgmt/ ) sort of like how the password resource does.

b) This could be fed into a "transform" block (resource) which performs all of the steps and modifications you mentioned above. Other interesting transform blocks could also be built and used elsewhere too. (Which is pretty cool.)

c) Finally it outputs to a file resource via send->recv which sets the contents.

d) These three components could then be optionally wrapped into a composite resource to make the whole use case simpler.

e) Lastly, I have an idea about how to automatically create composite resources by defining sub graph (dag's) inside an individual resource. It will be a while before I get to implement this unless someone wants to work on it.

To be clear, this third option is the most complex, but potentially the most fun and research/exploratory like if someone is interested, but I'd be perfectly happy to ACK the first option if someone wants this sooner rather than later.

HTH and thanks!

**Comment by mildred** [comment]
I just implemented the line editing feature in a separate lib: https://github.com/mildred/go-fileline

**Comment by purpleidea** [comment]
@mildred Cool! Want to implement (1) and import your lib for the functionality? You can either:

a) Keep your lib as is, assuming you'll be keeping git master stable.
or
b) Copy it into mgmt as a package there with the new resource.

I'd probably prefer (b) but am okay with (a) if you feel strongly about it.

Cheers!

**Comment by mildred** [comment]
Is it possible to have something like that, a loop between two resources:

- a file_line resource that uses RecvSend to receive the file content before it has changed from a file resource and generates an output

- a file resource that watches the file for changes and uses SendRecv to receive the processed content from the file_line resource

This makes a loop between two resources, I wonder how this can be solved. Perhaps we need something like 'facts'. The file resource would trigger when the file fact would tell it a change had been made, and would use output from file_line resource which itself would look for the file fact the actual content on disk for its unprocessed input.

Basically, this means resources as they are currently could be split in two. The watch part of the resource is a 'fact' that could be used independently.

**Comment by purpleidea** [comment]
@mildred The design I mentioned above is all still correct, but I understand the line of reasoning you're going down... Indeed it might make sense to have "fact" like things that can take an argument of a file path to read it in, but I currently left this out of the lang prototype for the moment.

Nevertheless, it's not as important a distinction because I think we're confusing the split puppet model where some things execute on the server and some on the client. We have no such problem since there is only one agent.

Don't worry about the loop, it's safe and converges.

**Comment by daenney** [comment]
I know this is a popular feature in other places, like Puppet, but it's also a rather common source of problems. Whichever way this gets implemented it's usually rather sensitive to the source that's being modified changing.

From my point of view, all we're doing is applying a patch to a config file, like the venerable `patch` allows you to do. Patch tends to also be less sensitive to things moving around a bit. Couldn't we implement this with the use of something like [go-diff](https://github.com/sergi/go-diff) instead, where the patch to apply could (optionally) be a [text/template](https://golang.org/pkg/text/template/) that is evaluated first?

**Comment by purpleidea** [comment]
We have a "line" resource now. If someone wants to add more features, please lmk.
